MediaBrowser.MediaEncoding/Encoder/MediaEncoder.cs

```diff
@@ -32,323 +28,288 @@ namespace MediaBrowser.MediaEncoding.Encoder
         }
 
-        public string EncoderLocationType
+        /// <summary>
+        /// Run at startup or if the user removes a Custom path from transcode page.
+        /// Sets global variables FFmpegPath and EncoderLocationType.
+        /// If startup options --ffprobe is given then FFprobePath is set too.
+        /// </summary>
+        public void Init()
         {
-            get
+            // 1) If given, use the --ffmpeg CLI switch
+            if (ValidatePathFFmpeg("From CLI Switch", StartupOptionFFmpegPath))
             {
-                if (_hasExternalEncoder)
-                {
-                    return "External";
-                }
-
-                if (string.IsNullOrWhiteSpace(FFMpegPath))
-                {
-                    return null;
-                }
-
-                if (IsSystemInstalledPath(FFMpegPath))
-                {
-                    return "System";
-                }
-
-                return "Custom";
+                _logger.LogInformation("FFmpeg: Using path from command line switch --ffmpeg");
+                EncoderLocationType = "External";
             }
-        }
 
-        private bool IsSystemInstalledPath(string path)
-        {
-            if (path.IndexOf("/", StringComparison.Ordinal) == -1 && path.IndexOf("\\", StringComparison.Ordinal) == -1)
+            // 2) Try Custom path stroed in config/encoding xml file under tag <EncoderAppPathCustom>
+            else if (ValidatePathFFmpeg("From Config File", ConfigurationManager.GetConfiguration<EncodingOptions>("encoding").EncoderAppPathCustom))
             {
-                return true;
+                _logger.LogInformation("FFmpeg: Using path from config/encoding.xml file");
+                EncoderLocationType = "Custom";
             }
 
-            return false;
-        }
-
-        public void Init()
-        {
-            InitPaths();
-
-            if (!string.IsNullOrWhiteSpace(FFMpegPath))
+            // 3) Search system $PATH environment variable for valid FFmpeg
+            else if (ValidatePathFFmpeg("From $PATH", ExistsOnSystemPath("ffmpeg")))
             {
-                var result = new EncoderValidator(_logger, _processFactory).Validate(FFMpegPath);
-
-                SetAvailableDecoders(result.decoders);
-                SetAvailableEncoders(result.encoders);
+                _logger.LogInformation("FFmpeg: Using system $PATH for FFmpeg");
+                EncoderLocationType = "System";
+            }
+            else
+            {
+                _logger.LogError("FFmpeg: No suitable executable found");
+                FFmpegPath = null;
+                EncoderLocationType = null;
             }
-        }
-
-        private void InitPaths()
-        {
-            ConfigureEncoderPaths();
 
-            if (_hasExternalEncoder)
+            // If given, use the --ffprobe CLI switch
+            if (ValidatePathFFprobe("CLI Switch", StartupOptionFFprobePath))
```

Comment:
Why `ffprobe` is handled differently from `ffmpeg`? I think there should be a parametrized method to search for `ff*` called twice (for `ffmpeg` and `ffprobe`)

Reply:
The idea is they always live together, so why repeat the discovery for the probe.  This is what the old code did.  What benefit in performing full search for probe please?

Reply:
Plus at the moment we don't store the probe path in the xml, and its not user configurable separately to the mpeg.

Reply:
Made some change to handle the mpeg and probe parts the same.